lib/matplotlib/legend_handler.py

```diff
@@ -672,17 +688,32 @@ def create_artists(self, legend, orig_handle,
         else:
             pad = self._pad * fontsize
 
-        if ndivide > 1:
-            width = (width - pad * (ndivide - 1)) / ndivide
+        if self._width_ratios is not None:
+            sumratios = sum(self._width_ratios)
+            widths = [(width - pad * (ndivide - 1)) * ratio / sumratios
+                      for ratio in self._width_ratios]
+        else:
+            widths = [(width - pad * (ndivide - 1)) / ndivide
+                      for _ in range(ndivide)]
+        widths_cycle = cycle(widths)
 
-        xds_cycle = cycle(xdescent - (width + pad) * np.arange(ndivide))
+        xds = [xdescent - (widths[-i-1] + pad) * i for i in range(ndivide)]
+        xds_cycle = cycle(xds)
 
         a_list = []
-        for handle1 in orig_handle:
-            handler = legend.get_legend_handler(handler_map, handle1)
-            _a_list = handler.create_artists(
-                legend, handle1,
-                next(xds_cycle), ydescent, width, height, fontsize, trans)
+        for i, handle1 in enumerate(orig_handle):
+            if self._handlers is not None:
+                handler = self._handlers[i]
+            else:
+                handler = legend.get_legend_handler(handler_map, handle1)
+
+            _a_list = handler.create_artists(legend, handle1,
+                                             six.next(xds_cycle),
```

Comment:
We don't need `six` anymore.  This will be py3 only....

Reply:
Can you explain (like I'm dumb) why we need widths to cycle here?  I'm not grokking the width_ratios and why they are needed.

Reply:
my fault for the `six` i merged with the other pr without seeking the py2->py3  changes

Reply:
this was pulled from #8292 and i played with it in order to find a reason to keep it.
basically, in the use case were a user would like both an arrow AND text in the legend, it would show up as 
![cycled_widths](https://user-images.githubusercontent.com/14252814/37044037-6182f982-2130-11e8-8dfd-96fa58537e3d.png)

and not as this:
![no_cycle_width](https://user-images.githubusercontent.com/14252814/37044047-6af1fe32-2130-11e8-8236-447a8a63e1bc.png)

and the "sections" that you were asking for further explanation for, are what overlap in the latter picture.
When the widths aren't cycled, the part of the legend where a text or an arrow would show up for the annotation would be drawn on top of one another.
So `width_ratios` was defined so that each arrow "section" and text "section" would be given their respective space in the legend box in the appropriate location, where their respective widths would be cycled in order to place them both together.

Reply:
@jklymak

Reply:
Ok this seems great.  but some of this should be in the doc string so it’s clear. Sections is a bit vague.

Reply:
Could be shorter. Ie width_ratios specifies the restive widths of a text/arrow legend annotation pair

Reply:
Gotcha 👍 
thanks again!
last thing left on my list is writing better tests :)